p2p/pre2p/raintree/network.go

```diff
@@ -0,0 +1,207 @@
+}
+
+func (n *rainTreeNetwork) networkBroadcastInternal(data []byte, level uint32, nonce uint64) error {
+	// This is handled either by the cleanup layer or redundancy layer
+	if level == 0 {
+		return nil
+	}
+
+	msg := &typesPre2P.RainTreeMessage{
+		Level: level,
+		Data:  data,
+		Nonce: nonce,
+	}
+	bz, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	if addr1, ok := n.getFirstTargetAddr(level); ok {
+		if err = n.networkSendInternal(bz, addr1); err != nil {
+			log.Println("Error sending to peer during broadcast: ", err)
+		}
+	}
+
+	if addr2, ok := n.getSecondTargetAddr(level); ok {
+		if err = n.networkSendInternal(bz, addr2); err != nil {
+			log.Println("Error sending to peer during broadcast: ", err)
+		}
+	}
+
+	if err = n.demote(msg); err != nil {
+		log.Println("Error demoting self during RainTree message propagation: ", err)
+	}
+
+	return nil
+}
+
+func (n *rainTreeNetwork) demote(rainTreeMsg *typesPre2P.RainTreeMessage) error {
+	if rainTreeMsg.Level > 0 {
+		if err := n.networkBroadcastInternal(rainTreeMsg.Data, rainTreeMsg.Level-1, rainTreeMsg.Nonce); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *rainTreeNetwork) NetworkSend(data []byte, address cryptoPocket.Address) error {
+	msg := &typesPre2P.RainTreeMessage{
+		Level: 0, // Direct send that does not need to be propagated
+		Data:  data,
+		Nonce: getNonce(),
+	}
+
+	bz, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	return n.networkSendInternal(bz, address)
+}
+
+func (n *rainTreeNetwork) networkSendInternal(data []byte, address cryptoPocket.Address) error {
+	// NOOP: Trying to send a message to self
+	if n.selfAddr.Equals(address) {
+		return nil
+	}
+
+	peer, ok := n.addrBookMap[address.String()]
+	if !ok {
+		return fmt.Errorf("address %s not found in addrBookMap", address.String())
+	}
+
+	if err := peer.Dialer.Write(data); err != nil {
+		log.Println("Error writing to peer during send: ", err)
+		return err
+	}
+
+	return nil
+}
+
+func (n *rainTreeNetwork) HandleNetworkData(data []byte) ([]byte, error) {
+	var rainTreeMsg typesPre2P.RainTreeMessage
+	if err := proto.Unmarshal(data, &rainTreeMsg); err != nil {
+		return nil, err
+	}
+
+	networkMessage := types.PocketEvent{}
+	if err := proto.Unmarshal(rainTreeMsg.Data, &networkMessage); err != nil {
+		log.Println("Error decoding network message: ", err)
+		return nil, err
+	}
+
+	// Continue RainTree propagation
+	if rainTreeMsg.Level > 0 {
+		if err := n.networkBroadcastInternal(rainTreeMsg.Data, rainTreeMsg.Level-1, rainTreeMsg.Nonce); err != nil {
+			return nil, err
+		}
+	}
+
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(rainTreeMsg.Nonce))
+	hash := cryptoPocket.SHA3Hash(b)
+	hashString := hex.EncodeToString(hash)
+	// Avoids this node from processing a messages / transactions is has already processed at the
+	// application layer. The logic above makes sure it is only propagated and returns.
+	if n.mempool.Contains(hashString) {
```

Comment:
I think this is good enough de-dupping, wdyt?

Reply:
Adding this TODO: 
```
// TODO(team): Add more tests to verify this is sufficient for deduping purposes.
```